# PA 3: Association Analysis - Apriori Algorithm

## Student Details 

Group Number :- 4

#### Group members name with ID :- 
Varun Adit Dhamudaran Sathish - 1001944885

Supreetha Darbha - 1002035242  

Subhash Duggirala - 1001965810






In [1]:
# Import your Libraries
import pandas as pd
from pandas import DataFrame
import csv
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import apriori
from apriori import getItemSetTransactionList
from apriori import returnItemsWithMinSupport
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import fpgrowth
from mlxtend.frequent_patterns import association_rules
from apriori import runApriori
from apriori import printResults
from apriori import dataFromFile
from apriori import joinSet
from apriori import subsets
from IPython.display import Image
# Above are the only libraries you can use. Do not make any changes above.
# -10 will be deducted, if you use additional libraries apart from these.

# - - - - - - - - - - - - - - - - - - - - - - - SOLUTION - - - - - - - - - - - - - - - - - - - - - - -

                                     ########## Task 1 ###########
                                     
--> Before you start, Please modify your dataset '(given dataset).csv' to look like the toyDS.csv.  

--> Export the modified data into new dataset named 'dataset.csv'

--> Read and print first 10 transactions of dataset.csv







In [2]:
########## Code for Task 1 #############
new_list=[]
with open("dataset_three.csv", "rt") as openfile, open('toyDS.csv', 'w') as f:
    for line in openfile:
        line=line.replace('{','').replace('}','').replace('\"','').replace('\'','').replace('\n','')
        new_list.append(line.rstrip().split(','))

final_list=[]
for row in new_list :
    row.sort()
    row = [item.replace('all- purpose','all-purpose') for item in row]
    row = [item.lstrip() for item in row]
    final_list.append(row)
    
    
        
with open('dataset.csv', 'w', newline='') as f:
    # using csv.writer method from CSV package
    write = csv.writer(f)
    write.writerows(final_list)
    

    
with open('dataset.csv') as file:
    for i in range(10):
        line = next(file).strip()
        print(line+'\n')


all-purpose,aluminum foil,beef,butter,dinner rolls,flour,ice cream,laundry detergent,lunch meat,mixes,pork,shampoo,soap,soda,vegetables,yogurt,sandwich bags

aluminum foil,cereals,cheeses,dishwashing liquid/detergent,hand soap,individual meals,laundry detergent,milk,mixes,shampoo,toilet paper,tortillas,vegetables,waffles,yogurt,sandwich bags

bagels,cereals,cheeses,dinner rolls,hand soap,ice cream,ketchup,laundry detergent,lunch meat,milk,pork,poultry,sandwich loaves,shampoo,soap,soda,spaghetti sauce,toilet paper,vegetables,eggs

all-purpose,juice,lunch meat,soda,toilet paper,cereals

all-purpose,dinner rolls,flour,hand soap,individual meals,milk,mixes,paper towels,pasta,pork,poultry,sandwich loaves,soda,spaghetti sauce,toilet paper,tortillas,vegetables,waffles,yogurt,eggs

all-purpose,aluminum foil,bagels,butter,cereals,coffee/tea,dishwashing liquid/detergent,ketchup,laundry detergent,milk,paper towels,pasta,poultry,shampoo,soap,spaghetti sauce,toilet paper,vegetables,eggs

all-purpos

                                  ######## Task 2 ########
    
--> In this task, You should be able to execute and print FPgrowth results using mlxtend.

--> "DO NOT USE ANY OTHER LIBRARY !!!".

--> Execute apriori algorithm atleast 3 times for different combinations of support and print the results for the dataset.

--> Please do explain your reasoning for each combinations.





In [3]:
######### Code for Task 2 #############

file = open('dataset.csv', 'r') 
list_values = file.readlines()
new_list=[]
for _ in list_values:
    new_list.append(_.strip())
file.close() 

final_list=[]
for _ in new_list:
    final_list.append(_.split(','))

te = TransactionEncoder()
te.fit(final_list)
te_ary=te.transform(final_list)

df=pd.DataFrame(te_ary, columns=te.columns_)

df

all-purpose  aluminum foil  bagels   beef  butter  cereals  cheeses  \
0            True           True   False   True    True    False    False   
1           False           True   False  False   False     True     True   
2           False          False    True  False   False     True     True   
3            True          False   False  False   False     True    False   
4            True          False   False  False   False    False    False   
...           ...            ...     ...    ...     ...      ...      ...   
1134         True          False   False   True   False     True     True   
1135        False          False   False  False   False     True     True   
1136        False          False    True   True   False    False    False   
1137         True          False   False   True   False    False     True   
1138        False          False   False  False   False    False    False   

      coffee/tea  dinner rolls  dishwashing liquid/detergent  ...  shampoo  \
0          False          True                         False  ...     True   
1          False         False                          True  ...     True   
2          False          True                         False  ...     True   
3          False         False                         False  ...    False   
4          False          True                         False  ...    False   
...          ...           ...                           ...  ...      ...   
1134        True          True                          True  ...     True   
1135        True          True                          True  ...    False   
1136       False          True                          True  ...     True   
1137       False         False                         False  ...    False   
1138       False         False                         False  ...     True   

       soap   soda  spaghetti sauce  sugar  toilet paper  tortillas  \
0      True   True            False  False         False      False   
1     False  False            False  False          True       True   
2      True   True             True  False          True      False   
3     False   True            False  False          True      False   
4     False   True             True  False          True       True   
...     ...    ...              ...    ...           ...        ...   
1134   True  False            False   True         False      False   
1135   True  False             True  False         False      False   
1136   True  False            False   True         False       True   
1137   True   True             True   True          True      False   
1138  False   True            False  False         False      False   

      vegetables  waffles  yogurt  
0           True    False    True  
1           True     True    True  
2           True    False   False  
3          False    False   False  
4           True     True    True  
...          ...      ...     ...  
1134       False    False   False  
1135        True    False   False  
1136        True    False    True  
1137        True     True    True  
1138        True    False   False  

[1139 rows x 38 columns]

In [4]:
print ('Case 1 : At least 30% support')
frequent_itemsets1 =fpgrowth(df, min_support=0.40, use_colnames=True)
frequent_itemsets1

Case 1 : At least 30% support


support      itemsets
0  0.739245  (vegetables)
1  0.421422     (poultry)

In [5]:
rules = association_rules(frequent_itemsets1, metric="confidence", min_threshold=0)
rules = rules.sort_values('confidence', ascending=True)
rules#.head()

Empty DataFrame
Columns: [antecedents, consequents, antecedent support, consequent support, support, confidence, lift, leverage, conviction]
Index: []

#### The above rules are generated for purchases with support of atleast 40%, and minimum confidence 0%. 

In this case, the support threshold is set to 40%, a frequent itemset is generated for the items that occur in at least 40% of all transactions in the database.

However, upon building association rules, we see that there is no association between the two most frequent items of the transactions.

In [6]:
print ('Case 2 : At least 10% support')
frequent_itemsets2=fpgrowth(df,min_support=0.1,use_colnames=True)
frequent_itemsets2

Case 2 : At least 10% support


support                            itemsets
0     0.739245                        (vegetables)
1     0.398595                         (ice cream)
2     0.395083                        (lunch meat)
3     0.390694                              (soda)
4     0.388938                      (dinner rolls)
...        ...                                 ...
1401  0.125549  (aluminum foil, sugar, vegetables)
1402  0.120281   (paper towels, sugar, vegetables)
1403  0.131694        (cereals, sugar, vegetables)
1404  0.113257  (sandwich bags, sugar, vegetables)
1405  0.117647          (sugar, vegetables, mixes)

[1406 rows x 2 columns]

In [7]:
rules = association_rules(frequent_itemsets2, metric="confidence", min_threshold=0.5)
rules = rules.sort_values('confidence', ascending=True)
rules#.head()

antecedents   consequents  \
527               (sandwich loaves, vegetables)     (poultry)   
15                               (dinner rolls)     (poultry)   
330  (dishwashing liquid/detergent, vegetables)        (eggs)   
388              (individual meals, vegetables)  (lunch meat)   
584                         (pasta, vegetables)        (eggs)   
..                                          ...           ...   
75                 (laundry detergent, cereals)  (vegetables)   
329        (dishwashing liquid/detergent, eggs)  (vegetables)   
21                         (dinner rolls, eggs)  (vegetables)   
36                               (yogurt, eggs)  (vegetables)   
541                     (sandwich loaves, eggs)  (vegetables)   

     antecedent support  consequent support   support  confidence      lift  \
527            0.282704            0.421422  0.141352    0.500000  1.186458   
15             0.388938            0.421422  0.194908    0.501129  1.189137   
330            0.306409            0.389816  0.153644    0.501433  1.286333   
388            0.285338            0.395083  0.143108    0.501538  1.269450   
584            0.286216            0.389816  0.143986    0.503067  1.290527   
..                  ...                 ...       ...         ...       ...   
75             0.169447            0.739245  0.151010    0.891192  1.205543   
329            0.171203            0.739245  0.153644    0.897436  1.213990   
21             0.173837            0.739245  0.156277    0.898990  1.216092   
36             0.174715            0.739245  0.157155    0.899497  1.216779   
541            0.134328            0.739245  0.121159    0.901961  1.220111   

     leverage  conviction  
527  0.022214    1.157155  
15   0.031001    1.159773  
330  0.034200    1.223876  
388  0.030376    1.213567  
584  0.032414    1.227902  
..        ...         ...  
75   0.025747    2.396463  
329  0.027083    2.542362  
21   0.027770    2.581475  
36   0.027998    2.594513  
541  0.021857    2.659701  

[715 rows x 9 columns]

#### The above rules are generated for purchases with support of atleast 10%, and minimum confidence 50%. 
In this case, the support threshold is set to 10%, a frequent itemset is generated for the items that occur in at least 40% of all transactions in the database.

Upon building association rules, we see that there is an association rule that gives us :
(sandwich loaves, vegetables) --> (poultry), confidence=0.5

This means that when a person does purchase sandwich loaves, and vegetables, then they will also purchase poultry with a confidence of 0.500 or 50% of the time.


In [8]:
print ('Case 3 : At least 6% support')
frequent_itemsets3=fpgrowth(df,min_support=0.06,use_colnames=True)
frequent_itemsets3

Case 3 : At least 6% support


support                                           itemsets
0      0.739245                                       (vegetables)
1      0.398595                                        (ice cream)
2      0.395083                                       (lunch meat)
3      0.390694                                             (soda)
4      0.388938                                     (dinner rolls)
...         ...                                                ...
10477  0.063213                               (eggs, sugar, mixes)
10478  0.062335                    (spaghetti sauce, sugar, mixes)
10479  0.069359                (poultry, sugar, vegetables, mixes)
10480  0.061457  (dishwashing liquid/detergent, sugar, vegetabl...
10481  0.061457           (dinner rolls, sugar, vegetables, mixes)

[10482 rows x 2 columns]

In [9]:
rules = association_rules(frequent_itemsets3, metric="confidence", min_threshold=0.3)
rules = rules.sort_values('confidence', ascending=False)
rules#.head()

antecedents  \
29565                   (sandwich loaves, poultry, eggs)   
17673      (poultry, dishwashing liquid/detergent, eggs)   
683                         (dinner rolls, eggs, yogurt)   
33600                (laundry detergent, cereals, pasta)   
1784   (aluminum foil, dishwashing liquid/detergent, ...   
...                                                  ...   
11161                                           (butter)   
10822                                           (butter)   
18190                                          (waffles)   
22598                                          (waffles)   
15001                                     (paper towels)   

                          consequents  antecedent support  consequent support  \
29565                    (vegetables)            0.070237            0.739245   
17673                    (vegetables)            0.086918            0.739245   
683                      (vegetables)            0.084284            0.739245   
33600                    (vegetables)            0.074627            0.739245   
1784                     (vegetables)            0.072871            0.739245   
...                               ...                 ...                 ...   
11161  (individual meals, vegetables)            0.367867            0.285338   
10822             (pasta, vegetables)            0.367867            0.286216   
18190              (milk, vegetables)            0.394205            0.295874   
22598         (hand soap, vegetables)            0.394205            0.270413   
15001              (vegetables, pork)            0.362599            0.270413   

        support  confidence      lift  leverage  conviction  
29565  0.067603    0.962500  1.302004  0.015681    6.953468  
17673  0.083406    0.959596  1.298076  0.019153    6.453687  
683    0.080773    0.958333  1.296368  0.018466    6.258121  
33600  0.071115    0.952941  1.289074  0.015947    5.541045  
1784   0.069359    0.951807  1.287540  0.015490    5.410667  
...         ...         ...       ...       ...         ...  
11161  0.110623    0.300716  1.053894  0.005657    1.021991  
10822  0.110623    0.300716  1.050661  0.005334    1.020736  
18190  0.118525    0.300668  1.016205  0.001890    1.006856  
22598  0.118525    0.300668  1.111886  0.011927    1.043263  
15001  0.108867    0.300242  1.110311  0.010816    1.042628  

[42118 rows x 9 columns]

#### The above rules are generated for purchases with support of atleast 6%, and minimum confidence 30%. 
In this case, the support threshold is set to 6%, a frequent itemset is generated for the items that occur in at least 6% of all transactions in the database.

Upon building association rules, we see that there is an association rule that gives us :
(paper towels) --> (pork, vegetables), confidence=0.300242

This means that when a person does purchase paper towels, then they will also purchasepork, and vegetables with a confidence of 0.300242 or about 30.02% of the time.


                                  ######## Task 3 ########
                                  
--> In this task, You should be able to execute and print apriory results by making use of apriory.py

--> "DO NOT USE ANY OTHER LIBRARY !!!".

--> Execute apriori algorithm atleast 3 times for different combinations of confidence and support and print the results for the dataset.

--> Please do explain your reasoning for each combinations.






In [10]:
########### Code for Task 3 ##############

print ('Task 3, Case:1') 

fileData = dataFromFile('dataset.csv')
items, rules = runApriori(fileData, 0.05, 0.96)

printResults(items, rules)

Task 3, Case:1

------------ITEMS-----------------
item: ('toilet paper', 'pork', 'fruits') , 0.050
item: ('flour', 'hand soap', 'bagels') , 0.050
item: ('hand soap', 'yogurt', 'dishwashing liquid/detergent') , 0.050
item: ('hand soap', 'tortillas', 'all-purpose') , 0.050
item: ('hand soap', 'pork', 'fruits') , 0.050
item: ('milk', 'pork', 'fruits') , 0.050
item: ('flour', 'juice', 'fruits') , 0.050
item: ('tortillas', 'juice', 'fruits') , 0.050
item: ('hand soap', 'cereals', 'pork') , 0.050
item: ('sandwich loaves', 'sugar', 'milk') , 0.050
item: ('hand soap', 'bagels', 'all-purpose') , 0.050
item: ('sandwich loaves', 'sugar', 'pork') , 0.050
item: ('hand soap', 'sandwich bags', 'individual meals') , 0.050
item: ('flour', 'beef', 'butter') , 0.050
item: ('lunch meat', 'pasta', 'pork') , 0.050
item: ('hand soap', 'yogurt', 'ice cream') , 0.050
item: ('hand soap', 'vegetables', 'butter', 'sandwich loaves') , 0.050
item: ('tortillas', 'milk', 'vegetables', 'sugar') , 0.050
item: ('soda',

#### The above rules are generated for purchases with support of atleast 5%, and minimum confidence 96%

One of the rules generated is :     ('sandwich loaves', 'dishwashing liquid/detergent', 'eggs') ==> ('vegetables',) , 0.968. 

This means that a person purchases sandwich loaves, dishwashing liquid/detergent, and eggs in 5% of the transactions.

And further, when a person does purchase sandwich loaves, dishwashing liquid/detergent, and eggs, then they will also purchase vegetables with a confidence of 0.968 or 96.8% of the time.

In [11]:
print ('Task 3, Case:2') 

fileData = dataFromFile('dataset.csv')

items, rules = runApriori(fileData, 0.07, 0.95)
printResults(items, rules)

Task 3, Case:2

------------ITEMS-----------------
item: ('cereals', 'soap', 'juice') , 0.070
item: ('spaghetti sauce', 'cheeses', 'ketchup') , 0.070
item: ('sandwich loaves', 'ketchup', 'mixes') , 0.070
item: ('tortillas', 'sandwich bags', 'ketchup') , 0.070
item: ('paper towels', 'dishwashing liquid/detergent', 'soap') , 0.070
item: ('beef', 'paper towels', 'soda') , 0.070
item: ('poultry', 'eggs', 'ketchup') , 0.070
item: ('dinner rolls', 'soap', 'spaghetti sauce') , 0.070
item: ('spaghetti sauce', 'milk', 'ketchup') , 0.070
item: ('laundry detergent', 'lunch meat', 'soap') , 0.070
item: ('poultry', 'cheeses', 'pork') , 0.070
item: ('hand soap', 'lunch meat', 'poultry') , 0.070
item: ('juice', 'pork', 'mixes') , 0.070
item: ('tortillas', 'sugar', 'individual meals') , 0.070
item: ('yogurt', 'sandwich bags', 'ice cream') , 0.070
item: ('cereals', 'tortillas', 'toilet paper') , 0.070
item: ('soap', 'milk', 'coffee/tea') , 0.070
item: ('lunch meat', 'soap', 'toilet paper') , 0.070
item

#### The above rules are generated for purchases with support of atleast 7%, and minimum confidence 95%. 
One of the rules generated is :     ('poultry', 'dishwashing liquid/detergent', 'eggs') ==> ('vegetables',) , 0.960. 

This means that a person purchases poultry, dishwashing liquid/detergent, and eggs in 7% of the transactions.

And further, when a person does purchase poultry, dishwashing liquid/detergent, and eggs, then they will also purchase vegetables with a confidence of 0.960 or 96% of the time.

In [12]:
print ('Task 3, Case:3') 

fileData = dataFromFile('dataset.csv')

items, rules = runApriori(fileData, 0.1, 0.9)
printResults(items, rules)

Task 3, Case:3

------------ITEMS-----------------
item: ('vegetables', 'pork', 'fruits') , 0.101
item: ('beef', 'vegetables', 'butter') , 0.101
item: ('hand soap', 'vegetables', 'butter') , 0.103
item: ('hand soap', 'vegetables', 'pork') , 0.104
item: ('juice', 'vegetables', 'fruits') , 0.105
item: ('hand soap', 'flour', 'vegetables') , 0.106
item: ('flour', 'spaghetti sauce', 'vegetables') , 0.106
item: ('shampoo', 'tortillas', 'vegetables') , 0.106
item: ('ice cream', 'vegetables', 'pork') , 0.106
item: ('sandwich loaves', 'vegetables', 'pork') , 0.106
item: ('flour', 'beef', 'vegetables') , 0.107
item: ('hand soap', 'vegetables', 'all-purpose') , 0.107
item: ('hand soap', 'paper towels', 'vegetables') , 0.108
item: ('paper towels', 'vegetables', 'pork') , 0.109
item: ('hand soap', 'bagels', 'vegetables') , 0.109
item: ('hand soap', 'sandwich bags', 'vegetables') , 0.109
item: ('toilet paper', 'vegetables', 'fruits') , 0.109
item: ('pork', 'vegetables', 'butter') , 0.109
item: ('tor

#### The above rules are generated for purchases with support of atleast 10%, and minimum confidence 90%. 
The only rule generated is :     ('sandwich loaves', 'eggs') ==> ('vegetables',) , 0.902. 

This means that a person purchases sandwich loaves, and eggs in 10% of the transactions.A

And further, when a person does purchases sandwich loaves, and eggs, then they will also purchase vegetables with a confidence of 0.901 or about 90% of the time.

## References
1. http://rasbt.github.io/mlxtend/user_guide/frequent_patterns/fpgrowth/


# Rubric
* [02 points] - Student Details 
* [08 points] - Comply with submission instructions 
* [20 points] - DataSet Preprocess for Apriory.py  
* [20 points] - Run Apriory.py
* [20 points] - DataSet Preprocess for  FPgrowth
* [30 points] - Run FPgrowth
 






